mm2src/coins/qrc20/swap_ops.rs

```diff
@@ -0,0 +1,917 @@
+        receiver_addr: H160,
+    ) -> Result<TransactionEnum, String> {
+        let allowance = try_s!(self.allowance(self.swap_contract_address).await);
+
+        let mut outputs = Vec::default();
+        // check if we should reset the allowance to 0 and raise this to the max available value (our balance)
+        if allowance < value {
+            let balance = try_s!(self.my_balance().compat().await);
+            let balance = try_s!(wei_from_big_decimal(&balance, self.utxo.decimals));
+            if allowance > U256::zero() {
+                // first reset the allowance to the 0
+                outputs.push(try_s!(self.approve_output(self.swap_contract_address, 0.into())));
+            }
+            // set the allowance from 0 to `balance` after the previous output will be executed
+            outputs.push(try_s!(self.approve_output(self.swap_contract_address, balance)));
+        }
+
+        // when this output is executed, the allowance will be sufficient already
+        outputs.push(try_s!(self.erc20_payment_output(
+            id,
+            value,
+            time_lock,
+            &secret_hash,
+            receiver_addr
+        )));
+
+        self.send_contract_calls(outputs).await
+    }
+
+    pub async fn spend_hash_time_locked_payment(
+        &self,
+        payment_tx: UtxoTx,
+        secret: Vec<u8>,
+    ) -> Result<TransactionEnum, String> {
+        let Erc20PaymentDetails {
+            swap_id, value, sender, ..
+        } = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
+
+        let status = try_s!(self.payment_status(swap_id.clone()).await);
+        if status != eth::PAYMENT_STATE_SENT.into() {
+            return ERR!("Payment state is not PAYMENT_STATE_SENT, got {}", status);
+        }
+
+        let spend_output = try_s!(self.receiver_spend_output(swap_id, value, secret, sender));
+        self.send_contract_calls(vec![spend_output]).await
+    }
+
+    pub async fn refund_hash_time_locked_payment(&self, payment_tx: UtxoTx) -> Result<TransactionEnum, String> {
+        let Erc20PaymentDetails {
+            swap_id,
+            value,
+            receiver,
+            secret_hash,
+            ..
+        } = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
+
+        let status = try_s!(self.payment_status(swap_id.clone()).await);
+        if status != eth::PAYMENT_STATE_SENT.into() {
+            return ERR!("Payment state is not PAYMENT_STATE_SENT, got {}", status);
+        }
+
+        let refund_output = try_s!(self.sender_refund_output(swap_id, value, secret_hash, receiver));
+        self.send_contract_calls(vec![refund_output]).await
+    }
+
+    pub async fn validate_payment(
+        &self,
+        payment_tx: UtxoTx,
+        time_lock: u32,
+        sender: H160,
+        secret_hash: Vec<u8>,
+        amount: BigDecimal,
+    ) -> Result<(), String> {
+        let erc20_payment = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
```

Comment:
I think that generating expected script_pubkey using `erc20_payment_output` and then looking for output containing this script_pubkey might be better for following reasons:
1. It will probably have less code.
2. You can get tx receipts here and find the receipt with output index found on previous step.
3. If `qrc20 payment script_pubkey` generation fn changes we are automatically required to perform changes here. For example if new argument is added to the end of `qrc20 payment script_pubkey` we can forget (or new developer might simply don't know that it's required) to add it to Erc20PaymentDetails and not check it here.

The only problem is other node can use unknown gas price and gas limit, this can be excluded from checking.

Reply:
Done as we discussed in Discord :)